src/app/wallet/proposals/proposal-details/proposal-details.component.ts

```diff
@@ -0,0 +1,152 @@
+    chart: {
+      type: 'pieChart',
+      height: 250,
+      width: 250,
+      x: (d) => { return d.description; },
+      y: (d) => { return d.voters; },
+      showLabels: false,
+      donut: true,
+      legend: {
+        margin: {
+          top: 5,
+          right: 35,
+          bottom: 5,
+          left: 0
+        }
+      },
+      color: ['#02E8B0', '#ec4b50', '#108cda', '#f1cc00', '#7e6c95'], // green, red, blue, yellow, violet
+      tooltip: {
+        enabled: true,
+        hideDelay: 0,
+        useInteractiveGuideline: false
+      }
+    }
+  };
+
+  public selectedOption: VoteOption;
+  public proposalResult: ProposalResult;
+  // @TODO create exact type.
+  public voteDetails: VoteDetails;
+  public aleradyVoted: boolean = false
+  destroyed: boolean = false;
+
+  constructor(
+    private dialog: MatDialog,
+    private proposalService: ProposalsService,
+    private snackbarService: SnackbarService,
+    private modelsService: ModalsHelperService
+  ) { }
+
+  ngOnInit() {
+    if (this.proposal) {
+      this.getProposalResult();
+    }
+  }
+
+  getVoteDetails(): void {
+    this.proposalService.get(this.proposal.hash)
+    .takeWhile(() => !this.destroyed)
+    .subscribe((voteDetail: VoteDetails) => {
+      this.voteDetails = voteDetail;
+      this.aleradyVoted = true;
+    }, (error) => {
+      this.log.d('you not yet?', error)
+    });
+  }
+
+  getProposalResult(): void {
+    this.proposalService.result(this.proposal.hash)
+      .takeWhile(() => !this.destroyed)
+      .subscribe((result: ProposalResult) => {
+        this.proposalResult = result;
+
+        // No need to call this.getVoteDetails() until proposal doesn't have any vote!!
+        if (this.proposalResult.totalVotes) {
+          this.getVoteDetails();
+        }
+      });
+  }
+
+  vote() {
+    const previousVote = this.voteDetails ? this.voteDetails.ProposalOption : null;
+    if (previousVote && previousVote.optionId === this.selectedOption.optionId) {
+      this.snackbarService.open(
+        `You already voted with option ${this.selectedOption.description} for proposal ${this.proposal.title}`,
+        'info'
+      );
+      return;
+    }
+
+    const dialog = this.dialog.open(ProposalVoteConfirmationComponent);
+    dialog.componentInstance.setData({
+      ... this.proposal,
+      selectedOption: this.selectedOption
+    }, () => this.vateConfirmed())
+  }
+
+  vateConfirmed(): void {
+    this.modelsService.unlock({timeout: 3}, (status) => this.callVote())
+  }
+
+  callVote(): void {
+    const params = [
+      this.proposal.hash,
+      this.selectedOption.optionId
+    ];
+    this.proposalService.vote(params).subscribe((response) => {
+      this.aleradyVoted = true;
+      // update graph data.
+      this.updateGraphData();
+
+      this.snackbarService.open(`Successfully Vote for ${this.proposal.title}`, 'info');
+    }, (error) => {
+      this.snackbarService.open(error.message, 'warn');
```

Comment:
Is the error message here sufficiently descriptive for a user to understand what went wrong?